File: 2021-07-21 (day3)/Husky/solution.js
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function (s) {
    let max_palindrome = s[0]
    for (let i = 0; i < s.length; i++) {
        let endIndex = s.length;
        console.log(s[i])

        if (endIndex - i < max_palindrome.length) {
            break
        }

        while (true) {
            const chars = s.slice(i, i + max_palindrome.length).split('').reverse().join('')
            endIndex = s.lastIndexOf(chars, endIndex - 1) + max_palindrome.length - 1
            console.log(chars, endIndex, max_palindrome)
            if (endIndex < 0 || endIndex - max_palindrome.length + i === i) {
                break
            }

            const chars_ = s.slice(i, endIndex + 1)
            const chars__ = chars_.split('').reverse().join('')

            console.log(chars_)
            if (chars_ === chars__ && chars_.length > max_palindrome.length) {
                max_palindrome = chars_
                console.log('123')
                break
            }
        }

        // console.log(endIndex)
    }

    return max_palindrome
};

// const test_cases = ["borcdubqiupahpwjizjjbkncelfazeqynfbrnzuvbhjnyvrlkdyfyjjxn  prfocrmisugizsgvhszooktdwhehojbrdbtgkiwkhpfjfcstwcajiuediebdhukqnroxbgtvottummbypokdljjvnthcbesoyigscekrqswdpalnjnhcnqrarxuufzzmkwizptyvjhpfidgmskuaggtpxqisdlyloznkarxofzeeyvteynluofuhbllyiszszbwnsglqjkignusarxsjvctpgiwnhdufmfpanfwxjwlmhgllzsmdpncbwnsbdtsvrjybynifywkulqnzprcxockbhrhvnwxrkvwumyieazclcviumvormynbryaziijpdinwatwqppamfiqfiojgwkvfzyxadyfjrgmtttvlgkqghgbcfhkigfojjkhskzenlpasyozcsuccdxhulcubsgomvcrbqwakrraesfifecmoztayrdjicypakrrneulfbqqxtrdelggedvloebqaztmfyfkhuonrognejxpesmwgnmnnnamvkxqvzrgzdqtvuhccryeowywmbixktnkqnwzlzfvloyqcageugmjqihyjxhssmhkfzxpzxmgpjgsduogfolnkkqizitbvvgrkczmojxnabruwwzxxqcevdwvtiigwckpxnnxxxdhxpgomncttjutrsvyrqcfwxhexhaguddkiesmfrqfjfdaqbkeqinwicphktffuwcazlmerdhraufbpcznbuipmymipxbsdhuesmcwnkdvilqbnkaglloswcpqzdggnhjdkkumfuzihilrpcvemwllicoqiugobhrwdxtoefynqmccamhdtxujfudbglmiwqklriolqfkknbmindxtljulnxhipsieyohnczddabrxzjmompbtnnxvivmoyfzfrxlufowtqzojfclmatthjtbhotdoheusnpirwicbtyrcuojsdmfcx"]; //
const test_cases = ["babad"] //, "cbbd", "a", "ac", "ccc", "cccc", "aacbcaa"

test_cases.forEach((c) => {
    palindromic = longestPalindrome(c)
    console.log(`${c}: \t${palindromic} `)
})

File: 2021-07-21 (day3)/Husky/solution.py
class Solution:
    def __init__(self):
        self.max_palindrome = ''

    def checkPalindrome(self, s: str) -> bool:
        return s == s[::-1]

    def replacePalindrome(self, s: str) -> bool:
        if len(s) > len(self.max_palindrome):
            self.max_palindrome = s

    def findNext(self, s: str, start: int, end: int) -> bool:
        palindromic = s[start:end + 1]
        if self.checkPalindrome(palindromic):
            self.replacePalindrome(palindromic)
            return False
        return True

    def longestPalindrome(self, s: str) -> str:
        self.max_palindrome = s[0]
        for i, char in enumerate(s):
            end_index = len(s)
            if end_index - i < len(self.max_palindrome):
                break

            while True:
                end_index = s.rfind(char, i + 1, end_index)
                if end_index < 0:
                    break

                if len(self.max_palindrome) > end_index - i + 1:
                    break

                next = self.findNext(s, i, end_index)
                if not next:
                    break

        return self.max_palindrome

if __name__ == '__main__':
    solution = Solution()

    test_cases = ["babad", "cbbd", "a", "ac", "ccc", "cccc", "aacbcaa"]

    for case in test_cases:
        palindromic = solution.longestPalindrome(case)
        print(f'{case}: \t{palindromic}')